ssh-pkcs11.c

```diff
@@ -525,17 +540,18 @@ pkcs11_fetch_keys_filter(struct pkcs11_provider *p, CK_ULONG slotidx,
 			    == NULL) {
 				error("d2i_X509 failed");
 			} else if ((evp = X509_get_pubkey(x509)) == NULL ||
-			    evp->type != EVP_PKEY_RSA ||
-			    evp->pkey.rsa == NULL) {
+			    EVP_PKEY_id(evp) != EVP_PKEY_RSA ||
+			    EVP_PKEY_get0_RSA(evp) == NULL) {
 				debug("X509_get_pubkey failed or no rsa");
-			} else if ((rsa = RSAPublicKey_dup(evp->pkey.rsa))
+			} else if ((rsa = RSAPublicKey_dup(EVP_PKEY_get0_RSA(evp)))
 			    == NULL) {
 				error("RSAPublicKey_dup");
 			}
 			if (x509)
 				X509_free(x509);
 		}
-		if (rsa && rsa->n && rsa->e &&
+		RSA_get0_key(rsa, &n, &e, NULL);
```

Comment:
This needs to be rewritten to something like the following, otherwise it will segfault for non-RSA keys (ECC on yubikey):
```
if (rsa && RSA_get0_key(rsa, &n, &e, NULL) && n && e &&
```
I just noticed it with the version I run on Fedora.

Reply:
nope ... the `RSA_get0_key()` is void function so it will not work this way. But clearly we can call this function only if we are sure the `rsa` is not NULL.

Reply:
> the RSA_get0_key() is void function so it will not work this way. But clearly we can call this function only if we are sure the rsa is not NULL

Just my 2¢, but if `RSA_get0_key` can't handle a NULL parameter, then it sounds like an OpenSSL bug. The [`RSA_get0_key` document](https://www.openssl.org/docs/man1.1.0/crypto/RSA_get0_key.html) does not state the parameter *cannot* be NULL; and it also states other parameters can be NULL.